### ISSUE TITLE
Feature: add Kafka transactions support

### ISSUE THREAD

**Issue by ManiMozaffar** [ag2ai/faststream#1157]
**Is your feature request related to a problem? Please describe.**
In batch processing, a group of messages is processed together. Transactions ensure that this entire batch is treated as a single unit of work. Although library support adding it to the producer config, but it doesn't support actualy using it.

**Describe the solution you'd like**
Some document and reworks regarding low level access to producer and consumer would be really nice in general. I think going forward this would be needed.

**Comment by Lancetnik** [comment]
@ManiMozaffar which broker are you talking?

**Comment by ManiMozaffar** [comment]
Kafka

**Comment by ManiMozaffar** [comment]
@Lancetnik either all messages should be produced, or none.

**Comment by davorrunje** [comment]
@ManiMozaffar we'll release a new Kafka broker based on Confluent lib this week. This is a good time to let us know what features you would like us to support.

**Comment by ManiMozaffar** [comment]
@davorrunje I haven't had a chance yet to deep dive into code, but on documentation would be really nice to give some idea about low level access to broker and other stuff, I think that would be nessecary because of many things like instrumenting the broker,  fixing weird bugs sometimes happening with brokers. E.x if broker dies and it was on transaction, it would go to a state where it can't produce a message transactionality so you'd have to restart the broker. (it loses the state of producer mapping to transactional id, or aka INVALID_PRODUCER_ID_MAPPING)

**Comment by RubenRibGarcia** [comment]
Hello once again! @Lancetnik I was reading a bit about the story for this feature ( specially a closed PR #1290 ) and I would like to talk a little bit about it's approach. That is, I know there is another feature on the Roadmap for RMQ to also support publish transaction and I think it would be best that both approaches followed a same pattern.

Focusing on Kafka... does it make sense that for KafkaBroker (both AioKafka and Confluent) have the `publish` decorator with a `transaction` argument ? The idea behind this is to abstract the way to handle transactions on both libs, but I'm also thinking that we should also provide some sort of annotation for developer to have more control over the transaction using explicit "begin", "commit", "rolback" methods.

**Comment by Lancetnik** [comment]
@RubenRibGarcia It is a much complex feature we image before and I am even not sure, that we can support it in FastStream API (in Kafka case at least)

The main idea now - is Depends-based transactional for RMQ (and Kafka probably)

```python
@broker.subscriber(...)
async def handler(
    msg,
    transaction: Transaction  # will be commited after handler reseived
): ...
```


Anyway, we will back to this discussion little bit further

**Comment by andreaimprovised** [comment]
Interested in this as well

**Comment by iliya-malecki** [comment]
@Lancetnik why not get inspired by FastAPI and `starlette.responses.StreamingResponse` and do something like

```python
@broker.subscriber(...)
@broker.publisher("a")
@broker.publisher("b")
@broker.publisher("c")
async def handler(
    msg,
): 
  return Transaction({
    "a":message_a,
    "b":message_b,
    "c":message_c,
  })
```

I think it hits the main design considerations, as im seeing/feeling them:
- topic connectivity decoupled from everything else, to keep dependency graph analysis possible by third-party tools
- returning from a function semantically is the act of sending
- it logically works nicely with the lifecycle and Middlewares, although i admit i dont immediately see how we could turn the multiple per-topic registrations into a single lifecycle event. Im mildly scared to take that challenge on so lets talk before i start

**Comment by Lancetnik** [comment]
@iliya-malecki I don't see any advantages in your suggested design. Also, it doesn't match `StreamingResponse` API you mention - it returns async generator, but you suggest to return multiple messages in a turn.

I would like to stay with my idea:

```python
@broker.subscriber(...)
async def handler(msg, transaction: Transaction):
    await transaction.add("message1", topic="a")
    await transaction.add("message2", topic="b")
```

It seems much closer to original transaction logic

**Comment by iliya-malecki** [comment]
- I meant it matches StreamingResponse in the sense that the processor function returns a magic instance that special cases the behaviour or processing the outputs in a hopefully clean manner
- removing the producer marker from the decorator list will introduce coupling that will make reasoning about the pubsub dependency graph impossible, which felt like an important detail of the decorator approach 
- the semantics I want to keep is return == send

**Comment by Lancetnik** [comment]
> * removing the producer marker from the decorator list will introduce coupling that will make reasoning about the pubsub dependency graph impossible, which felt like an important detail of the decorator approach

It has a sense, but I don't like suggested API. I try to find a balance between a simple API and powerful features. Powerful things don't always work out on a simple API, sometimes you have to sacrifice something. We already have such examples.

What if we provide an ability to pass `transaction` to `publish`?

```python
publisher = broker.publisher("a")

@broker.subscriber(...)
async def handler(msg, transaction: Transaction):
    await publisher.publish("message1", transaction=transaction)
    await broker.publish("message2", topic="b", transaction=transaction)
```

**Comment by iliya-malecki** [comment]
@Lancetnik I agree that powerful Apis can require tradeoffs, but can you elaborate on why you don't like the API design I'm proposing? Maybe something like `@broker.publisher.transaction("a", "b", "c")` would be nicer as the decorator? I'm sure there is an idea you will enjoy, I dont want to get stuck on one approach - I have plenty :) 
P.s. I am also interested in implementing something

**Comment by Lancetnik** [comment]
Nope, it also doesn't work for me `@broker.publisher.transaction("a", "b", "c")` not clear at all and introduces a new abstraction we didn't have before. We should introduce a new API for single feature.

**Comment by Lancetnik** [comment]
Also, I don't like to limit transaction usage to decorator-API only. Decorators are too simple and can't cover a lot of cases.

Like dynamic publishing:

```python
@broker.subscriber(...)
async def handler(reply_to: Annotated[str, Header()]):
    await broker.publish("reply", reply_to)
```

Or conditional publishing:

```python
@broker.subscriber(...)
async def handler():
    if condition:
        await broker.publish("reply", "a")
    else:
        await broker.publish("reply", "b")
```

New API for transactions should cover all possible scenarios. This reason I would like to stay with my current UoW-like design

**Comment by iliya-malecki** [comment]
Fair. I agree that dynamic topic choice can be a legitimate thing to do, and should be supported. In my approach, it turns Transaction ({"topic":message}) into a Publish({"topic":message}), with a separate place for a transaction switch. I am not sure if I love the verbosity of this. It has a chance to play nicely with the middleware, but at a cost of being a major pain.

Tldr I can agree that with the current constraints retrofitting a richer publishing approach is challenging
